### PR TITLE
remove usage of Java bytecode frontend

### PR DESCRIPTION
**Pull request by kroening**
Do not merge yet -- we'll give test-gen time to switch to the jbmc repository.

### CODE REVIEW

**Comment by chrisr-diffblue** [comment]
I've pushed what I think should be a fix to get the CMake builds passing again (I can now build CMake builds locally). I have to say it's a bit of a hack, mostly to work around the fact the the CMake builds are currently using file globing - but also because I'm not entirely sure what the full intent of this PR is - are you intending to entirely move the java_bytecode source out of the repo? Or still keep it in the repo, but not build it by default? If the code is going to stay in the repo then it may be a better idea to refactor the directory structure a bit so that all the java_bytecode + it's unit/regression tests are in their own sub-directory. This would make it then possible to have CMake cleanly ignore the java_bytecode stuff.

File: regression/Makefile
Comment (tautschnig) [inline regression/Makefile]: This seems wrong. If there are Java-only bits then there should be a goto-diff-java directory instead.

File: unit/Makefile
Comment (tautschnig) [inline unit/Makefile:18]: If those three depend on Java then the implementation ought to move to the `java_bytecode` folder.

File: unit/Makefile
Comment (tautschnig) [inline unit/Makefile]: Why is this being removed?

File: unit/Makefile
Comment (tautschnig) [inline unit/Makefile]: Change the test, don't remove it.

File: unit/Makefile
Comment (tautschnig) [inline unit/Makefile]: Either the test is inappropriate or it shouldn't be removed.

**Review by tautschnig — CHANGES_REQUESTED** [review]
1) Documentation (at the very least: COMPILING.md) needs to be updated as well.
2) The changes to regression and unit tests seem wrong as indicated.

File: regression/Makefile
Comment (chrisr-diffblue) [reply regression/Makefile]: It looks like it's 11 out of 17 tests in goto-diff that use java, but I agree, it seems wrong to be disabling non-java tests in this PR! I've replicated this particular change in the CMakeFile fix up commit, but that doesn't mean I agree with that change :-)

**Review by chrisr-diffblue — CHANGES_REQUESTED** [review]
While I've 'fixed up' the cmake builds so that they pass CI, I strongly agree with @tautschnig that the unit/regression test changes are wrong - there's useful non-java tests being disabled. It would also help reviewing if there was a clear statement as to what the strategy is for the java front-end now - if the java_bytecode stuff is going to be removed, then why is that removal not just part of this PR? Although it would be a 'bigger' PR, it would probably also be cleaner.

File: src/goto-cc/CMakeLists.txt
Comment (tautschnig) [inline src/goto-cc/CMakeLists.txt]: This is a surprising change.

File: unit/CMakeLists.txt
Comment (tautschnig) [inline unit/CMakeLists.txt]: I thought it was being removed?

File: unit/CMakeLists.txt
Comment (tautschnig) [inline unit/CMakeLists.txt]: I am confused, unless GitHub is just displaying the commits in the wrong order?

File: unit/Makefile
Comment (tautschnig) [inline unit/Makefile]: miniz is being removed as well.

**Review by tautschnig — COMMENTED** [review]
Some comments below; I think the update to `COMPILING.md` is still missing.

File: unit/CMakeLists.txt
Comment (chrisr-diffblue) [reply unit/CMakeLists.txt]: This is an annoying quirk of how our CMake files work... It's basically collecting an exclusion list here, so this change expands the exclusion list to include all files in the java_bytecode directory tree (i.e. all the test cases), rather than *just* the java testing utils.

File: unit/CMakeLists.txt
Comment (chrisr-diffblue) [reply unit/CMakeLists.txt]: This change basically picks up the expanded exclusion list from the previous patch hunk.

File: unit/Makefile
Comment (chrisr-diffblue) [reply unit/Makefile]: I believe that is because the only use-case for miniz is for reading .jar files?

File: unit/CMakeLists.txt
Comment (tautschnig) [reply unit/CMakeLists.txt]: Ah ... so a future commit will actually remove those files and these lines as well?

File: src/goto-cc/CMakeLists.txt
Comment (chrisr-diffblue) [reply src/goto-cc/CMakeLists.txt]: I was a bit surprised too :-) but it seemed to be needed to avoid link-time errors. I think the json dependency was previously being fulfilled transitively via one of the now-removed java libraries.  If you think this really is a problem I can happily dig a bit deeper as to what/where that dependency is coming from.

File: unit/Makefile
Comment (tautschnig) [reply unit/Makefile]: Yes. With the changes in this PR the `miniz$(OBJEXT)` file would never get built anymore.

File: unit/CMakeLists.txt
Comment (chrisr-diffblue) [reply unit/CMakeLists.txt]: Yep, once the various java directories are actually removed, yes.

File: unit/Makefile
Comment (peterschrammel) [reply unit/Makefile]: Yes, miniz should be removed too.

File: src/goto-cc/CMakeLists.txt
Comment (tautschnig) [reply src/goto-cc/CMakeLists.txt]: Not a big problem, just surprising.

Comment (chrisr-diffblue) [reply src/goto-cc/CMakeLists.txt]: BTW, this change to the CMakeLists.txt basically just brings it into line with the un-touched Makefile for this directory, which has `../json/json$(LIBEXT)` as an `OBJ` to be linked - so I think change is just fixing up what was a broken CMake config...

**Comment by tautschnig** [comment]
My remaining concerns are:
1) Missing update to `COMPILING.md`
2) `miniz` still being built/required in unit tests.

**Comment by chrisr-diffblue** [comment]
I've added two more commits that should address @tautschnig's concerns. I've left these as additional commits at the moment, just to ease reviewing, but they should be squashed into @kroening's commit before merging.

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
I maintain that moving Java into a separate repository is in my opinion not the right approach, but as the decision appears to be set the changes made here are consistent thus I've given my approval to this PR.

**Review by chrisr-diffblue — APPROVED** [review]
I've not been involved in the decision to remove Java into a separate repository, but assuming thats a done deal - and with the *big* caveat that I'm now rather too close to this PR... I'm approving.

**Review by peterschrammel — CHANGES_REQUESTED** [review]
I'm blocking this until the move has finished.

**Review by peterschrammel — CHANGES_REQUESTED** [review]
This is passing now on all platforms with all build systems. I've seen passing builds of TG on Friday. 

I don't think we have to wait for any other PRs to be merged before, because they can now be more easily rebased (since everything remains in the same repo).

I would appreciate a second eye pair to sanity check the PR before merging, though. There have been quite some changes since the approving reviews.

The addition of jdiff and janalyzer suffers from the same shortcomings as https://github.com/diffblue/cbmc/pull/2144, but cannot be postponed. I suggest to apply the required simplifications in a later step as soon as we have a solution for 2144.

File: jbmc/README.md
Comment (kroening) [inline jbmc/README.md]: The executable is called jbmc, not symex.

File: jbmc/src/CMakeLists.txt
Comment (kroening) [inline jbmc/src/CMakeLists.txt]: Does jbmc need bison/flex at all?

File: jbmc/src/janalyzer/Makefile
Comment (kroening) [inline jbmc/src/janalyzer/Makefile]: The assembler front-end isn't needed.

File: jbmc/src/janalyzer/janalyzer_parse_options.h
Comment (kroening) [inline jbmc/src/janalyzer/janalyzer_parse_options.h]: No need for D:I: etc. in janalyzer.

File: jbmc/src/janalyzer/janalyzer_parse_options.h
Comment (kroening) [inline jbmc/src/janalyzer/janalyzer_parse_options.h]: No need for --gcc option.

File: jbmc/src/jbmc/CMakeLists.txt
Comment (kroening) [inline jbmc/src/jbmc/CMakeLists.txt]: No need for assembler.

File: jbmc/src/jbmc/CMakeLists.txt
Comment (kroening) [inline jbmc/src/jbmc/CMakeLists.txt]: No need for cpp.

File: jbmc/src/jdiff/CMakeLists.txt
Comment (kroening) [inline jbmc/src/jdiff/CMakeLists.txt]: Same remove assembler, cpp

**Comment by tautschnig** [comment]
Should `janalyzer` and `jdiff` really even exist? I'm aware that `goto-analyzer` currently does support C programs as input, but I don't think it actually should. There's a front-end tool (`goto-cc`) to build goto binaries and working with those ought to be completely sufficient. Now of course I don't actually know about the Java story. Since lazy loading doesn't seem to make as much sense for `goto-analyzer` (or `goto-diff`), shouldn't there just be an equivalent of `goto-cc` for Java? As far as I can tell there is no need for `goto-analyzer` (or `goto-diff`) to produce any language-specific output, and consequently these could safely be used without any linking to C/Java/other front-end languages?

**Comment by peterschrammel** [comment]
`janalyzer` and `jdiff` should only provide a Java-friendly command line interface (eventually); they shouldn't provide any core algorithms. This is currently suboptimal (I essentially copied them) and needs to be improved to avoid code duplication in *_parse_options. I agree that there should be goto-cc for Java, but that's orthogonal. You could argue in the same way that CBMC need not accept any .c files. The current problem that is solved here is that at the time of merging this PR we need a goto-diff variant that accepts jar files.

**Comment by tautschnig** [comment]
> `janalyzer` and `jdiff` should only provide a Java-friendly command line interface (eventually); they shouldn't provide any core algorithms. This is currently suboptimal (I essentially copied them) and needs to be improved to avoid code duplication in *_parse_options. I agree that there should be goto-cc for Java, but that's orthogonal. You could argue in the same way that CBMC need not accept any .c files. The current problem that is solved here is that at the time of merging this PR we need a goto-diff variant that accepts jar files.

I accept that you might have an immediate business need for a working `goto-diff` for jar files, hence a copy&paste `jdiff` is a short-term workaround.

In the medium term: It's not clear to me what constitutes a "Java-friendly command line interface?" As I said, I don't see why any of those tools should have anything to do with the input language at all, and thus I don't really know why such a command-line interface would be needed? Of course any of that only applies when we have a goto-cc equivalent for Java, which in turn I thus do not see as orthogonal but rather as a necessary dependency to enable language-independence of goto-analyzer/goto-diff (and ideally also goto-instrument, but that requires a lot more cleanup).

Edit: yes, I would indeed argue that CBMC need to accept any .c files. It's just that a lot of people out there have gotten used to using it in that way, so it requires a slightly longer transition period.

**Review by peterschrammel — APPROVED** [review]
TG build is passing. I'm fine to get this merged now.
@kroening, @smowton, @thk123, any objections?

**Pull request merged**
